src/main/scala/org/broadinstitute/dsde/workbench/sam/dataAccess/PostgresGroupDAO.scala

```
@@ -0,0 +1,324 @@
+  *  15636 |     7795 |                 | userid
+  *  15637 |     7798 |                 | userid
+  *  15638 |     7801 |            7798 |
+  *  15639 |     7801 |            7799 |
+  *
+  * testdb=# select * from sam_group_member_flat;
+  *    id   | group_id | member_group_id | member_user_id | group_membership_path
+  * --------+----------+-----------------+----------------+-----------------------
+  *  345985 |     7795 |                 | userid         | {7795}
+  *  345986 |     7798 |                 | userid         | {7798}
+  *  345987 |     7801 |            7798 |                | {7801}
+  *  345988 |     7801 |            7799 |                | {7801}
+  *  345989 |     7801 |                 | userid         | {7801,7798}
+  *
+  * It is crucial that all group updates are in serializable transactions to avoid race conditions when
+  * concurrent modifications are made affecting the same group structure.
+  */
+trait PostgresGroupDAO {
+  protected def insertGroupMembers(groupId: GroupPK, members: Set[WorkbenchSubject])(implicit session: DBSession): Int =
+  {
+    val memberGroupPKs = queryForGroupPKs(members)
+    val memberUserIds = collectUserIds(members)
+
+    insertGroupMemberPKs(groupId, memberGroupPKs, memberUserIds)
+  }
+
+  protected def collectUserIds(members: Set[WorkbenchSubject]): List[WorkbenchUserId] = {
+    members.collect {
+      case userId: WorkbenchUserId => userId
+    }.toList
+  }
+
+  protected def insertGroupMemberPKs(groupId: GroupPK, memberGroupPKs: List[GroupPK], memberUserIds: List[WorkbenchUserId])(implicit session: DBSession): Int = {
+    if (memberGroupPKs.isEmpty && memberUserIds.isEmpty) {
+      0
+    } else {
+      val insertCount = Try {
+        insertGroupMembersIntoHierarchical(groupId, memberGroupPKs, memberUserIds)
+      }.recover {
+        case duplicateException: PSQLException if duplicateException.getSQLState == PSQLStateExtensions.UNIQUE_VIOLATION => 0
```

I always get antsy when swallowing exceptions.  If inserting multiple rows and one of them causes problems, it might be nice to have logged that error so we know which row is the culprit.

Furthermore, what triggers transaction failures?  Does the fact that a transaction was thrown, handled, and swallowed force any applicable transactions to fail?  Or do we need to keep bubbling up the exception and or manually trigger the failure/rollback?

---

would adding `on conflict do nothing` to the query instead make you feel better?

---

actually `on conflict do nothing` is functionally correct, as implemented it is probably a bug